### PR TITLE
Fix to #13642 - Comparison of nullable result of functions

### PR DESCRIPTION
**Pull request by maumar**
Problem was that we were not applying clr null semantics when comparing result of functions, i.e. if the function results in null and we compared it to another null value, we would treat them as non-equal. According to clr null semantics they should be treated as the same.

Fix is to expand IsNullExpression builder to handle functions with nullable arguments.

f(a, b) == null <=> a == null || b == null, unless function is COALESCE (in that case both arguments have to be null), so we can just test arguments not the entire function expression.
Also added small optimization in PredicateReductionExpressionOptimizer that removes redundant IS NULL terms.

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/Query/NullSemanticsQuerySqlServerTest.cs [hunk 1]
```diff
@@ -1194,6 +1194,56 @@ public override void Projecting_nullable_bool_with_coalesce_nested()
...
+FROM [Entities1] AS [e]
+WHERE ((CHARINDEX(N'oo', [e].[NullableStringA]) - 1) = [e].[NullableIntA]) OR ([e].[NullableStringA] IS NULL AND [e].[NullableIntA] IS NULL)",
+                //
+                @"SELECT [e].[Id]
+FROM [Entities1] AS [e]
+WHERE ((CHARINDEX(N'ar', [e].[NullableStringA]) - 1) = [e].[NullableIntA]) OR ([e].[NullableStringA] IS NULL AND [e].[NullableIntA] IS NULL)",
+                //
+                @"SELECT [e].[Id]
+FROM [Entities1] AS [e]
+WHERE (((CHARINDEX(N'oo', [e].[NullableStringA]) - 1) <> [e].[NullableIntB]) OR ([e].[NullableStringA] IS NULL OR [e].[NullableIntB] IS NULL)) AND ([e].[NullableStringA] IS NOT NULL OR [e].[NullableIntB] IS NOT NULL)");
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/NullSemanticsQuerySqlServerTest.cs]: this translation looks weird but is actually correct, consider the truth table breakdown in RelationalNullsExpandingVisitor.ExpandNullableNotEqualNullable

File: test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs [hunk 2]
```diff
@@ -1239,7 +1239,7 @@ public override async Task Select_null_propagation_negative6(bool isAsync)
                 @"SELECT CASE
     WHEN [g].[LeaderNickname] IS NOT NULL
     THEN CASE
-        WHEN CAST(LEN([g].[LeaderNickname]) AS int) <> CAST(LEN([g].[LeaderNickname]) AS int)
+        WHEN ((CAST(LEN([g].[LeaderNickname]) AS int) <> CAST(LEN([g].[LeaderNickname]) AS int)) OR [g].[LeaderNickname] IS NULL) AND [g].[LeaderNickname] IS NOT NULL
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs]: this should be optimized out to nothing, will do in another PR when I add more complex optimizations

File: src/EFCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs [hunk 3]
```diff
@@ -43,33 +43,36 @@ protected override Expression VisitConstant(ConstantExpression constantExpressio
...
+                || binaryExpression.NodeType == ExpressionType.OrElse)
             {
-                // a ?? b == null <-> a == null && b == null
+                AddToResult(new IsNullExpression(binaryExpression));
 
-                var current = ResultExpression;
-                ResultExpression = null;
-                Visit(binaryExpression.Left);
-                var left = ResultExpression;
+                return binaryExpression;
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs]: Use else and remove this return statement

File: src/EFCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs [hunk 4]
```diff
@@ -83,8 +86,41 @@ protected override Expression VisitExtension(Expression extensionExpression)
             if (extensionExpression is NullableExpression nullableExpression)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs]: Convert to switch statement

File: src/EFCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs [hunk 5]
```diff
@@ -83,8 +86,41 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
             {
                 AddToResult(new IsNullExpression(nullableExpression.Operand));
+
+                return extensionExpression;
+            }
+
+            if (extensionExpression is ExplicitCastExpression explicitCastExpression)
+            {
+                return Visit(explicitCastExpression.Operand);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/Internal/IsNullExpressionBuildingVisitor.cs]: Is there a specific reason we are not calling into base? Calling base would cover this case automatically and may be for other expression types too

File: src/EFCore.Relational/Query/ExpressionVisitors/Internal/PredicateReductionExpressionOptimizer.cs [hunk 6]
```diff
@@ -45,6 +46,11 @@ protected override Expression VisitBinary(BinaryExpression node)
                         // a && false => false
                         return (bool)constantRight.Value ? newLeft : newRight;
                     }
+
+                    if (ExpressionEqualityComparer.Instance.Equals(newLeft, newRight))
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/Internal/PredicateReductionExpressionOptimizer.cs]: Move it outside of if to make it DRY

Comment (smitpatel) [reply src/EFCore.Relational/Query/ExpressionVisitors/Internal/PredicateReductionExpressionOptimizer.cs]: Also move it to very first

---
In reply to: [227952182](https://github.com/aspnet/EntityFramework/pull/13734#discussion_r227952182) [](ancestors = 227952182)

File: src/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs [hunk 7]
```diff
@@ -7055,6 +7055,26 @@ select new
...
+        {
+            return AssertQuery<CogTag>(
+                isAsync,
+                ts => ts.Where(t => t.Note.Substring(0, t.Gear.SquadId) == t.GearNickName),
+                ts => ts.Where(t => Maybe(t.Gear, () => t.Note.Substring(0, t.Gear.SquadId)) == t.GearNickName));
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Null_semantics_is_correctly_applied_for_function_comparisons_that_take_arguments_from_optional_navigation_complex(bool isAsync)
```
Comment (smitpatel) [inline src/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs]: So complex case works but basic fails :trollface:

File: test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs]: Could we do simple optimization that a != a => false

---
In reply to: [227620517](https://github.com/aspnet/EntityFramework/pull/13734#discussion_r227620517) [](ancestors = 227620517)

File: src/EFCore.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs [hunk 8]
```diff
@@ -275,7 +275,7 @@ private static Expression ExpandNullableEqualNullable(Expression left, Expressio
 
         // ?a != ?b -> [(a != b) || (a == null || b == null)] && (a != null || b != null)]
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs]: Not sure how you should write it but better way to represent this is
Since, `a != b` both of them being null means false, so `(a != b) || (a is null && b is not null) || (a is not null && b is null)`

Comment (maumar) [reply src/EFCore.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs]: when both a and b are null this would return null, we want to have only 1s or 0s as the result of rewrite, since its the "full" expansion - getting rid of nulls so we don't need to worry about null semantics further up the pipeline.
